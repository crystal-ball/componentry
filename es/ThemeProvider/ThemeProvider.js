import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function () { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import { Component } from 'react';
import { number, shape } from 'prop-types';
/**
 * The `<ThemeProvider>` is a shorthand for setting context values that can be used
 * for changing the default configuration values used by Componentry components.
 * The passed theme configurations are namespaced to prevent collisions.
 *
 * See the theme propTypes for available configurations.
 *
 * Note that we don't actually set any context here for default, b/c using the
 * ThemeProvider is entirely optional. Any theme defaults need to be handled by
 * the component (or /utils/theme when components). This way when a user doesn't
 * use the ThemeProvider there is still defaults.
 * @export
 * @class ThemeProvider
 * @extends {Component}
 */

var ThemeProvider = /*#__PURE__*/function (_Component) {
  _inherits(ThemeProvider, _Component);

  var _super = _createSuper(ThemeProvider);

  function ThemeProvider() {
    _classCallCheck(this, ThemeProvider);

    return _super.apply(this, arguments);
  }

  _createClass(ThemeProvider, [{
    key: "getChildContext",

    /**
     * Return a library namespace for theme configurations to prevent name collisions
     */
    value: function getChildContext() {
      return {
        THEME: _objectSpread({}, this.props.theme)
      };
    }
  }, {
    key: "render",
    value: function render() {
      return this.props.children;
    }
  }]);

  return ThemeProvider;
}(Component);

_defineProperty(ThemeProvider, "childContextTypes", {
  THEME: shape({
    transitionDuration: number // Component configurations not declared, explicit declaration not necessary

  })
});

export { ThemeProvider as default };