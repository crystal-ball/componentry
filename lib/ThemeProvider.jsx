import { Component } from 'react';
import { element, number, shape, string } from 'prop-types';

/**
 * The `<ThemeProvider>` is a shorthand for setting context values that can be used
 * for changing the default configuration values used by Componentry components.
 * The passed theme configurations are namespaced to prevent collisions.
 *
 * See the theme propTypes for available configurations.
 *
 * @export
 * @class ThemeProvider
 * @extends {Component}
 */
export default class ThemeProvider extends Component {
  static childContextTypes = {
    componentry_visibilityTransitionLength: number,
    componentry_svgDefinitionsFilePath: string
  };

  static propTypes = {
    children: element.isRequired,
    theme: shape({
      visibilityTransitionLength: number, // Duration of visibility transitions
      svgDefinitionsFilePath: string // File path used in SVG use href
    })
  };

  static defaultProps = {
    theme: {}
  };

  /**
   * Namespace all theme configurations with `componentry_` to prevent any possible
   * name collisions in consuming apps. Return value to set context for all library
   * components.
   */
  getChildContext() {
    const { theme } = this.props;
    const nameSpacedTheme = {};

    Object.keys(theme).forEach(config => {
      nameSpacedTheme[`componentry_${config}`] = theme[config];
    });

    return nameSpacedTheme;
  }

  render() {
    return this.props.children;
  }
}
