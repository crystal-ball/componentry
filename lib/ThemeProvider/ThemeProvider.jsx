import { Component } from 'react'
import { element, number, shape, string } from 'prop-types'

/**
 * The `<ThemeProvider>` is a shorthand for setting context values that can be used
 * for changing the default configuration values used by Componentry components.
 * The passed theme configurations are namespaced to prevent collisions.
 *
 * See the theme propTypes for available configurations.
 *
 * @export
 * @class ThemeProvider
 * @extends {Component}
 */
export default class ThemeProvider extends Component {
  static childContextTypes = {
    COMPONENTRY_THEME: shape({
      defaultButtonColor: string,
      svgDefinitionsFilePath: string,
      visibilityTransitionLength: number
    })
  }

  static propTypes = {
    children: element.isRequired,
    theme: shape({
      defaultButtonColor: string, // Theme color to default to
      svgDefinitionsFilePath: string, // File path used in SVG use href
      visibilityTransitionLength: number // Duration of visibility transitions
    }).isRequired
  }

  /**
   * Return a library namespace for theme configurations to prevent name collisions
   */
  getChildContext() {
    return { COMPONENTRY_THEME: { ...this.props.theme } }
  }

  render() {
    return this.props.children
  }
}
