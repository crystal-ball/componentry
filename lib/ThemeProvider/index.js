// @flow
import { Component } from 'react'
import type { Node } from 'react'
import { number, shape, string } from 'prop-types'

/**
 * The `<ThemeProvider>` is a shorthand for setting context values that can be used
 * for changing the default configuration values used by Componentry components.
 * The passed theme configurations are namespaced to prevent collisions.
 *
 * See the theme propTypes for available configurations.
 *
 * Note that we don't actually set any context here for default, b/c using the
 * ThemeProvider is entirely optional. Any theme defaults need to be handled by
 * the component (or /utils/theme when components). This way when a user doesn't
 * use the ThemeProvider there is still defaults.
 * @export
 * @class ThemeProvider
 * @extends {Component}
 */

type Props = {
  children: Node,
  theme: {
    defaultButtonColor: string, // Theme color to default to
    transitionDuration: number // Duration of visibility transitions
  }
}

export default class ThemeProvider extends Component<Props> {
  static childContextTypes = {
    THEME: shape({
      defaultButtonColor: string,
      transitionDuration: number
    })
  }

  /**
   * Return a library namespace for theme configurations to prevent name collisions
   */
  getChildContext() {
    return { THEME: { ...this.props.theme } }
  }

  render() {
    return this.props.children
  }
}
