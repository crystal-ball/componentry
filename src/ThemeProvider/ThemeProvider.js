// @flow
import { Component, createContext } from 'react'
import type { Node } from 'react'
import { number, shape } from 'prop-types'

export const themeDefaults = { transitionDuration: 300 }

/**
 * The `<ThemeProvider>` is a shorthand for setting context values that can be used
 * for changing the default configuration values used by Componentry components.
 * The passed theme configurations are namespaced to prevent collisions.
 *
 * See the theme propTypes for available configurations.
 *
 * Note that we don't actually set any context here for default, b/c using the
 * ThemeProvider is entirely optional. Any theme defaults need to be handled by
 * the component (or /utils/theme when components). This way when a user doesn't
 * use the ThemeProvider there is still defaults.
 * @export
 * @class ThemeProvider
 * @extends {Component}
 */

const ThemeContext = createContext(themeDefaults)
export default ThemeContext

// LEGACY
// ---------------------------------------------------------------------------

export type Theme = {
  /** Default duration for theme visibility transitions */
  transitionDuration: number,
  /**
   * All components can have default behaviors configured using a context object for
   * that component. The values are spread into each component instance.
   */
  [string]: { [string]: any },
}

type Props = {
  children: Node,
  theme: Theme,
}

export class LegacyThemeProvider extends Component<Props> {
  static childContextTypes = {
    THEME: shape({
      transitionDuration: number,
      // Component configurations not declared, explicit declaration not necessary
    }),
  }

  /**
   * Return a library namespace for theme configurations to prevent name collisions
   */
  getChildContext() {
    return { THEME: { ...this.props.theme } }
  }

  render() {
    return this.props.children
  }
}
