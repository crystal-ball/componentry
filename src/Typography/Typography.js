import elem from '../elem-factory'
import { useTheme } from '../Theme/Theme'

const defaultVariantsElements = {
  'heading-1': 'h1',
  'heading-2': 'h2',
  'heading-3': 'h3',
  'body': 'p',
  'code': 'code',
  'small': 'small',
}

/**
 * [Typography component üìù](https://componentry.design/components/typography)
 */
export default function Typography(props) {
  const { variant = 'body', inline = false, variantsElements = {}, ...rest } = {
    ...useTheme('Typography'),
    ...props,
  }

  return elem({
    as: inline
      ? 'span'
      : variantsElements[variant] || defaultVariantsElements[variant] || 'p',
    elemClassName: inline ? null : Typography.classesPrefix + variant,
    ...rest,
  })
}
Typography.classesPrefix = '‚ú®'
Typography.displayName = '‚ú®Typography'
