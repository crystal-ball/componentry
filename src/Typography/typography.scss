/* componentry/src/Typography/typography */

$typography: () !default;

$typography-base: (
  // Component classNames prefix
    prefix: $component-prefix,
  // A class is generated for each variant
    variants:
    (
      // --- Heading 1 styles
        heading-1:
        (
          font-size: 3rem,
          margin-bottom: 0,
        ),
      // --- Heading 2 styles
        heading-2:
        (
          font-size: 2rem,
          margin-bottom: 0,
        ),
      // --- Heading 3 styles
        heading-3:
        (
          font-size: 1.5rem,
          margin-bottom: 0,
        ),
      // --- Body styles
        body:
        (
          font-size: 1rem,
          // Set spacing between multiple paragraphs using sibling selector and
            // margin-top.
            '& + &':
            (
              margin-top: 1rem,
            ),
        )
    )
);

$merged-typography: map-assign($typography-base, $typography);

@each $variant, $styles in get($merged-typography, variants) {
  @include style-rule-from-map('.#{get($merged-typography, prefix)}#{$variant}', $styles);
}
