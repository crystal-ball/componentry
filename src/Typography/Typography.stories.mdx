import { Meta, Props } from '@storybook/addon-docs/blocks'

import Typography from './Typography'
import Block from '../Block/Block'

<Meta title='Components/Typography' />

<DocsTitle title='Typography' />

## Headings

<Typography variant='heading-1'>Heading 1</Typography>
<Typography variant='heading-2'>Heading 2</Typography>
<Typography variant='heading-3'>Heading 3</Typography>

## Body

<Typography variant='body'>
  Life finds a way. Life finds a way. God help us, we're in the hands of engineers. Life
  finds a way. Must go faster. Hey, you know how I'm, like, always trying to save the
  planet? Here's my chance. Eventually, you do plan to have dinosaurs on your dinosaur
  tour, right?
</Typography>
<Typography variant='body'>
  Life finds a way. Life finds a way. God help us, we're in the hands of engineers. Life
  finds a way. Must go faster. Hey, you know how I'm, like, always trying to save the
  planet? Here's my chance. Eventually, you do plan to have dinosaurs on your dinosaur
  tour, right?
</Typography>

## Small

<Typography variant='small'>Small copy</Typography>

## Code

<Typography variant='code'>Code element</Typography>

#### Font sizes

<Typography variant='body' fontSize='sm'>
  Small size body.
</Typography>
<Typography variant='body' fontSize='md'>
  Medium size body.
</Typography>
<Typography variant='body' fontSize='lg'>
  Large size body.
</Typography>
