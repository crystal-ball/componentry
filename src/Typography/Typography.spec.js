import React from 'react'
import { render } from '@testing-library/react'

import Typography from './Typography'
import ThemeProvider from '../Theme/Theme'
import elementTests from '../../test/element-tests'

describe('<Typography/>', () => {
  // Basic library element test suite
  elementTests(Typography)
})

describe('Typography', () => {
  test('When variantsElements is set in theme, then variants are used for Typogrpahy', () => {
    const { container } = render(
      <ThemeProvider theme={{ Typography: { variantsElements: { rad: 'section' } } }}>
        <Typography variant='rad'>Componentry</Typography>
      </ThemeProvider>,
    )
    expect(container.firstChild).toContainHTML(
      '<section class="âœ¨rad">Componentry</section>',
    )
  })

  test('When inline is truthy, then a span without a variant class is rendered', () => {
    const { container } = render(<Typography inline>span content</Typography>)
    expect(container.firstChild).toContainHTML('<span class="">span content</span>')
  })
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Typography /> snapshots', () => {
  it('renders correctly', () => {
    const { container } = render(<Typography>Componentry</Typography>)
    expect(container.firstChild).toMatchSnapshot()
  })
})
