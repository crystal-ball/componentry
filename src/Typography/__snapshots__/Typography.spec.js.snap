// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Typography /> snapshots renders correctly 1`] = `
<p
  class="âœ¨body"
>
  Componentry
</p>
`;
